File: problems/0236/solution.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution:
    def lowestCommonAncestor(
        self, root: "TreeNode", p: "TreeNode", q: "TreeNode"
    ) -> "TreeNode":
        def dfsPostOrder(node: "TreeNode", p: "TreeNode", q: "TreeNode"):
            if not node:
                return None
            left = dfsPostOrder(node.left, p, q)
            right = dfsPostOrder(node.right, p, q)
            if node.val == p.val or node.val == q.val:
                return node

            if left and right:
                return node
            return left if left else right

        return dfsPostOrder(root, p, q)

File: problems/0236/solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */

class Solution {
 private:
  TreeNode* dfsPostOrder(TreeNode* node, TreeNode* p, TreeNode* q) {
    if (node == nullptr) return nullptr;

    TreeNode* left = dfsPostOrder(node->left, p, q);
    TreeNode* right = dfsPostOrder(node->right, p, q);

    if (node->val == p->val || node->val == q->val) {
      return node;
    }
    if (left != nullptr && right != nullptr) {
      return node;
    }
    return left ? left : right;
  }

 public:
  TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
    return dfsPostOrder(root, p, q);
  }
};